### PR TITLE
Added pooch registries for data files

### LINKED ISSUE #4026: Decide how to bundle `.pickle` files from COMSOL solution results data
`.pickle` files are binary files and not inherently secure. They therefore should be best included elsewhere in PyBaMM's packaging infrastructure when publishing to common distribution channels like PyPI owing to security reasons in the case where these binaries are modified with malicious intent and make it to the `develop` branch or a PyPI release where PyBaMM can be downloaded from (binary or non-binary). This means they should neither be included in the source distribution (controlled by `MANIFEST.in`), nor in the wheels (in the case where the wheel is built from the sdist). Binary platform-specific wheels for PyBaMM cannot be built from sdists right now and have to be built in-tree (#3651, #3564), and require restructuring the build system configuration at least, if not rewriting what files are copied into the package structure – since `pybamm/input/comsol_results/` is a namespace package (via `setuptools.command.install_data`).

We have the following options:

1. Store in JSON, CSV, or TXT file formats (non-binary) instead of `.pickle`
2. Use [`pooch`](https://learn.scientific-python.org/development/patterns/data-files/) to reproducibly download these files when needed and adjust PyBaMM's public API plus example notebooks for this, while moving these files to a separate repository in the `pybamm-team` organisation, and marking the repository as an archive to prevent tampering.
3. Use non-binary storage solutions from `xarray` or `zarr` because they can handle multi-dimensional data

Out of these, point 2 requires an internet connection and those wishing to load COMSOL results from these files would need access to the internet to download them, and therefore should be the last option to consider.

<hr>

## Task list, as of 10/05/2024

There are two major things to be done:

- [x] Convert pickle files to non-binary files (done in #4064)
- [ ] Use https://github.com/fatiando/pooch as an optional (`[dev]`) dependency to fetch these files plus adjust for the change here
    - [x] Set up a new repository: https://github.com/pybamm-team/pybamm-data
        - [ ] Add a README and some documentation
        - [ ] Copy the files and delete them from the PyBaMM main (this) repository (and additionally store checksums in a text file somewhere), and publish a GitHub Release there to store them perhaps (shorter URL)
    - [ ] Refer to the Scientific Python guide: https://learn.scientific-python.org/development/patterns/data-files/#downloading-larger-files-on-demand
    - [ ] Use `pytest`'s conditional markers or regular markers to skip tests where access to the internet is not available (adjust `noxfile.py`, test-related files, etc.)
    - [ ] Ensure files are not copied into the sdist (control `MANIFEST.in` and `setuptools`-specific package data)
    - [ ] Unsure about this point – maybe provide a public API to fetch these files (with the URLs kept hardcoded)?

### PR DESCRIPTION
**Pull request by santacodes**
# Description

- Added pooch as a [dev] dependency and added a file to create and store data file registries for [pybamm-data](https://github.com/pybamm-team/pybamm-data). 
- Added dataloader method to fetch data files from [pybamm-data](https://github.com/pybamm-team/pybamm-data) and store it under local cache folder and import it in notebooks wherever needed with specified paths.
- Removed data files from local repo under `pybamm/input`
- Data files under `pybamm/input/parameters` have not been removed or included in the pooch registries.
- Added documentation, tests and an example notebook for `pybamm.DataLoader` class. 

Fixes #4026 

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [x] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [x] All tests pass: `$ python run-tests.py --all` (or `$ nox -s tests`)
- [x] The documentation builds: `$ python run-tests.py --doctest` (or `$ nox -s doctests`)

You can run integration tests, unit tests, and doctests together at once, using `$ python run-tests.py --quick` (or `$ nox -s quick`).

## Further checks:

- [x] Code is commented, particularly in hard-to-understand areas
- [x] Tests added that prove fix is effective or that feature works

### CODE REVIEW

File: pybamm/pybammdata.py
Comment (agriyakhetarpal) [inline pybamm/pybammdata.py:1]: Could you rename the file to `pybamm_data.py`?

File: pybamm/pybammdata.py
Comment (agriyakhetarpal) [inline pybamm/pybammdata.py:5]: This class should include a docstring with Google-style docstring parameters to explain what it is about and a usage example or two. You should add the associated `.rst` files as well for it to render on the Sphinx documentation

File: pybamm/pybammdata.py
Comment (agriyakhetarpal) [inline pybamm/pybammdata.py:10]: Could you add a comment that the data under `pybamm/input/parameters/` has not been included at this time?

File: pybamm/pybammdata.py
Comment (agriyakhetarpal) [inline pybamm/pybammdata.py:43]:
```suggestion
    def get_data(self, filename: str):
```

File: pybamm/pybammdata.py
Comment (agriyakhetarpal) [inline pybamm/pybammdata.py:46]: This will most likely fail on Windows where the path separator is `\` rather than `/`. Could you construct and return a `Path` object instead (or otherwise, use `os.normpath` on this string)?

File: pybamm/pybammdata.py
Comment (agriyakhetarpal) [inline pybamm/pybammdata.py:48]: These lines need a test. A few simple test cases with dummy file names should work plus a few cases with actual files. You would need to add a marker to check if internet access is available, and skip the tests if not running under a network.

File: pyproject.toml
Comment (agriyakhetarpal) [inline pyproject.toml:114]: Could you make this a required dependency? Also, could you update the installation guide in the documentation about this – we have a table there that lists all of our dependencies and their minimum versions. It might also be good to add a minimum version here (one that is available for Python 3.9+ on both conda-forge and PyPI).

File: pybamm/pybammdata.py
Comment (santacodes) [reply pybamm/pybammdata.py:1]: sure!

File: pybamm/pybammdata.py
Comment (santacodes) [reply pybamm/pybammdata.py:10]: added this to the PR description

File: pybamm/pybammdata.py
Comment (santacodes) [reply pybamm/pybammdata.py:46]: `self.path` refers to the `pooch.os_cache` method which returns a path object, so I think it should be good.

Comment (agriyakhetarpal) [reply pybamm/pybammdata.py:46]: No, I meant the filename after `self.path`, which is still separated with `/`. This should be normalised. I would suggest using `pathlib`'s `Path` object, you can use `/` that way.

Comment (santacodes) [reply pybamm/pybammdata.py]: resolved in the latest commit

File: pybamm/pybammdata.py
Comment (santacodes) [reply pybamm/pybammdata.py]: I have added test cases with one dummy file and one randomly selected files from the registry

File: pybamm/pybamm_data.py
Comment (agriyakhetarpal) [inline pybamm/pybamm_data.py:123]:
```suggestion


    def show_registry(self, checksum: bool = False):
```

(style guide): it is generally better to leave out two newlines between methods.

File: pybamm/pybamm_data.py
Comment (agriyakhetarpal) [inline pybamm/pybamm_data.py:105]:
```suggestion


    def get_data(self, filename: str):
```

File: pybamm/pybamm_data.py
Comment (agriyakhetarpal) [inline pybamm/pybamm_data.py:29]:
```suggestion
        COMSOL Results
        ---------------
```

File: pybamm/pybamm_data.py
Comment (agriyakhetarpal) [inline pybamm/pybamm_data.py:40]: Could you add the relevant papers here (and similarly for other data files under the other headings where applicable)? They should be present in the `CITATIONS.bib` file. If they are not, you can get the BibTeX reference for them from Google Scholar or similar and add them to the file. Then, you can reference them in the docstring with something like

```rest

:footcite:t:`AuthorYear`

```

File: pybamm/pybamm_data.py
Comment (agriyakhetarpal) [inline pybamm/pybamm_data.py:19]:
```suggestion
       The folder locations are defined by the `platformdirs <https://platformdirs.readthedocs.io/en/latest/>`_ package using the `user_cache_dir <https://platformdirs.readthedocs.io/en/latest/api.html#cache-directory>`_ function.
```

I would suggest skipping these, it directs too much emphasis on users who do not have to deal with this information when using the `DataLoader` class, and therefore we can keep this abstracted away from them.

**Review by agriyakhetarpal — COMMENTED** [review]
Thanks for this, @santacodes! Could you fix the coverage by adding additional tests for the `get_data` method (to check the `NameError` that you raised), and for the `checksum` option in the `show_registry` method? @valentinsulzer, could you review this as well?

Besides this, I will recommend building the source distribution and the wheels on your fork by triggering the manual job we have available, and locally using `python -m build` to ensure that we aren't missing out on any files and that the sizes of the distributions are alright.

File: pybamm/pybamm_data.py
Comment (santacodes) [reply pybamm/pybamm_data.py:40]: Sure!

File: tests/unit/test_parameters/test_current_functions.py
Comment (santacodes) [reply tests/unit/test_parameters/test_current_functions.py:25]: Yeah oops I forgot to add those marks in other tests

File: tests/unit/test_pybamm_data.py
Comment (santacodes) [reply tests/unit/test_pybamm_data.py:4]: I think it is a required dependency of pooch

Comment (santacodes) [reply tests/unit/test_pybamm_data.py:4]: https://www.fatiando.org/pooch/latest/install.html

Comment (agriyakhetarpal) [reply tests/unit/test_pybamm_data.py:4]: Makes sense. I am not sure if we should include it as well – explicit is better than implicit so I would say we should, but happy to hear from others as well

**Comment by santacodes** [comment]
Ran the Build and publish package CI to build whls on my branch, and the sizes have been reduced by almost 15-20MB for POSIX and Windows distributions - https://github.com/santacodes/PyBaMM/actions/runs/9162618687.

File: tests/unit/test_pybamm_data.py
Comment (agriyakhetarpal) [reply tests/unit/test_pybamm_data.py]: Here, I would say we should use something from the standard library, like

```python
import socket

host = socket.gethostbyname("www.github.com")
conn = socket.create_connection((host, 80), 2)
conn.close()
```

inside a try-except block rather than importing a transitive dependency like `requests` that we do not explicitly mark as ours. This should be reasonable and diverts the question altogether, plus it helps us in case `pooch` ever decides to drop `requests` and aims to download files using plain-old `urllib` (not that it seems it will, though).

Comment (santacodes) [reply tests/unit/test_pybamm_data.py]: Sure that would work, I'll do that then!

File: pybamm/pybamm_data.py
Comment (agriyakhetarpal) [inline pybamm/pybamm_data.py:149]:
```suggestion
        Prints the names all the files present in the registry along with their checksums (optional).
```

File: pybamm/pybamm_data.py
Comment (agriyakhetarpal) [inline pybamm/pybamm_data.py:142]:
```suggestion

        Returns
        -------
        POSIX/Windows Path to the data file.
```

Also, could you add a proper return type for this?

File: pybamm/pybamm_data.py
Comment (agriyakhetarpal) [inline pybamm/pybamm_data.py:158]: I am not sure how I feel about a method that returns different data types based on the input argument, do we have a better way to do this?

File: docs/source/user_guide/installation/index.rst
Comment (santacodes) [reply docs/source/user_guide/installation/index.rst:72]: sure I fiexed it, I don't remember indenting it that way maybe the linter did it.

File: pybamm/pybamm_data.py
Comment (santacodes) [reply pybamm/pybamm_data.py:158]: I don't feel there might be a practical application for the return data types, it usually just prints the the registry which is the intention of that method and returns what it prints (to assert it in pytest unit tests).

File: pybamm/pybamm_data.py
Comment (santacodes) [reply pybamm/pybamm_data.py:15]: I think we might be well off removing the specifications when they are already mentioned in the `get_data()` method, and creating an `intersphinx` mapping for pooch and `platfordirs`. Will add it in the latest commit.

File: pybamm/pybamm_data.py
Comment (agriyakhetarpal) [reply pybamm/pybamm_data.py:158]: We could create a separate function `show_registry_by_checksum()` that returns a dictionary, while this one can return a list. I agree that there is not much of a practical application for this and I don't think end users need to view the checksums for any of the files (because they have been present inside the PyBaMM repository before this PR).

Another option could be to implement something like `get_data("US06.csv").checksum` that can the checksum for an individual file as a `str`, and move away from this one.

Comment (santacodes) [reply pybamm/pybamm_data.py]: I decided to remove the checksum parameter overall as it doesn't have any practical application, rather if a user wants to check the checksum of a particular file they could access the `files` attribute of the `data_loader` class which would give them a `dict` of all files and their checksums. I also updated its execution in the example notebook.

**Review by agriyakhetarpal — APPROVED** [review]
Looks good to me now, thanks, @santacodes! I'll wait for some additional thoughts on this before we merge.

cc @cringeyburger @prady0t – this PR has removed (most of) the PyBaMM data files from the PyBaMM source code. Testing the built wheels using `CIBW_TEST_COMMAND` should now be a bit easier. I feel that the `pytest --pyargs pybamm` invocation would be better than a `pybamm.test()` one – that way we would not need to move the `tests/` directory inside `pybamm/`

File: pyproject.toml
Comment (arjxn-py) [reply pyproject.toml:46]: No worries on keeping it as an optional dependency too, but can you specify which previously existent file? Feel free to resolve this conversation further.

File: docs/source/examples/notebooks/models/Validating_mechanical_models_Enertech_DFN.ipynb
Comment (arjxn-py) [reply docs/source/examples/notebooks/models/Validating_mechanical_models_Enertech_DFN.ipynb:1]: > this serves as a personal reminder to switch to MyST-NB

+1 for this as far as it is not a major breaking change, feel free to open an issue for the same or let me know to open it.

**Review by arjxn-py — APPROVED** [review]
I've no further concerns with this one, good work @santacodes.
@agriyakhetarpal are we good to merge this or we still seek +1 review?

File: pyproject.toml
Comment (agriyakhetarpal) [reply pyproject.toml:46]: Ah, for example, earlier to this PR one could do

```
drive_cycle = pd.read_csv(
    "pybamm/input/drive_cycles/US06.csv", comment="#", header=None
).to_numpy()
```

in a drive cycle simulation to retrieve a NumPy array from the CSV-converted `DataFrame` object – but, after this PR, one needs to use `data_loader.get_data("US06.csv")` to access it. We have added an entry about this in the CHANGELOG, which is fine, but we would also like people to receive functionalities to work with these files without having to install any additional optional dependencies such as `pybamm[data]` after they try to upgrade their workflows to v24.5. 

(keeping it optional would lead to another error about `pooch` not being installed, which could be disorienting to unsuspecting users who are not reading the CHANGELOG as often)

**Comment by agriyakhetarpal** [comment]
I think this is good to go on my end, but I just had a question about skipping tests without networks – @santacodes, the `pytest` decorator + `no_internet_connection()` we have will prevent the tests from failing, but this is just for the unit and the integration tests. Is there a way we can cleanly extend this behaviour for the example scripts and notebooks which make use of the data files, in specific? It's not a big deal though, because we expect that most people have internet access nowadays somehow or the other.

**Comment by santacodes** [comment]
> I think this is good to go on my end, but I just had a question about skipping tests without networks – @santacodes, the `pytest` decorator + `no_internet_connection()` we have will prevent the tests from failing, but this is just for the unit and the integration tests. Is there a way we can cleanly extend this behaviour for the example scripts and notebooks which make use of the data files, in specific? It's not a big deal though, because we expect that most people have internet access nowadays somehow or the other.

For notebooks and scripts, I am thinking we could add an error-handling method to raise an Error, pooch raises `socket.gaierror` by default when there is no network connection as it cannot create a socket connection with the upstream URL.
The second option would be to check for the internet connection before actually fetching data files from upstream and warning the user about network connectivity. 
I think both can be implemented, but I am not sure which would be more optimal for scripts and notebooks specifically.

**Comment by agriyakhetarpal** [comment]
> For notebooks and scripts, I am thinking we could add an error-handling method to raise an Error, pooch raises `socket.gaierror` by default when there is no network connection as it cannot create a socket connection with the upstream URL. The second option would be to check for the internet connection before actually fetching data files from upstream and warning the user about network connectivity. I think both can be implemented, but I am not sure which would be more optimal for scripts and notebooks specifically.

I thought of that but both of these options should ideally stay outside and abstracted away from the content inside the notebooks and scripts, which isn't quite viable and seems over-engineered... I opine that we should skip them for now. We can help out on a per-user basis if someone tries to use `nox -s scripts` or `nox -s examples` without an internet connection. Another reason why I say this is because we do not include any of the example scripts or notebooks inside the sdist or the wheel, which means that people have to come to GitHub or browse the online documentation to access them anyway, and accessing GitHub requires access to an internet connection.

File: tests/unit/test_pybamm_data.py
Comment (kratman) [inline tests/unit/test_pybamm_data.py:18]: I agree with codacy here, random in tests is not great. Can we just check the first file instead?
```suggestion
def test_fetch():
    test_file = list(pybamm.DataLoader().files.keys())[0]
    return data_loader.get_data(test_file).is_file()
```

File: tests/unit/test_pybamm_data.py
Comment (kratman) [inline tests/unit/test_pybamm_data.py:3]: With the other change, this should be removed
```suggestion
```

**Review by kratman — CHANGES_REQUESTED** [review]
Minor changes to the tests

**Review by kratman — COMMENTED** [review]
For any other reviewers: I checked the link failure. It should start passing once this PR is merged. The link checker goes with the stable documentation, so it cannot see the new page

**Comment by agriyakhetarpal** [comment]
Maybe we should refer to the reST-based heading for PyBaMM Data instead of adding an HTTP(S) link to our documentation? Otherwise, this failing link will keep failing for all the release candidates until we have the v24.5 release out – the link is pointing to `stable`, not `latest`. @santacodes, could you make this change in the notebook?

**Comment by agriyakhetarpal** [comment]
Changing to `../api/pybamm_data.rst` won't help – could you add a heading under the PyBaMM Data page and use reST syntax to link to it? Something under [this SO post](https://stackoverflow.com/questions/18632781/how-to-make-an-internal-link-to-a-heading-in-sphinx-restructuredtext-without-cre) should help

**Review by kratman — CHANGES_REQUESTED** [review]
Marking @agriyakhetarpal's last comment as a change request to keep this from being merged until it is fixed:

> Changing to `../api/pybamm_data.rst` won't help – could you add a heading under the PyBaMM Data page and use reST syntax to link to it? Something under [this SO post](https://stackoverflow.com/questions/18632781/how-to-make-an-internal-link-to-a-heading-in-sphinx-restructuredtext-without-cre) should help

**Comment by santacodes** [comment]
> Changing to `../api/pybamm_data.rst` won't help – could you add a heading under the PyBaMM Data page and use reST syntax to link to it? Something under [this SO post](https://stackoverflow.com/questions/18632781/how-to-make-an-internal-link-to-a-heading-in-sphinx-restructuredtext-without-cre) should help

I tried referencing headers, but sphinx doesn't seem to render it as a link no matter what, even when I reference the `pybamm_data.rst` to reference the documentation header as they are not under the same directory. It only works by referencing the notebook's path like in my last commit, or by referencing `docs/source/api/pybamm_data.rst:#pybamm.DataLoader`. They both mean the same thing, so I will leave it up to you to decide further.

**Comment by agriyakhetarpal** [comment]
I think this should be okay, then, thanks for the explanation. Not sure if there is a better way to link them unless we switch to MyST completely for the codebase, which is something for a later time

**Pull request merged**
